Content.Client/UserInterface/Controls/RadialMenu.cs

```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
         parent.TryToMoveToNewLayer(TargetLayer);
     }
 
+    /// <inheritdoc />
+    protected override bool HasPoint(Vector2 point)
+    {
+        var cX = -Position.X + Parent!.Width / 2;
+        var cY = -Position.Y + Parent.Width / 2;
+
+        var dist = Math.Sqrt(Math.Pow(point.X - cX, 2) + Math.Pow(point.Y - cY, 2));
+        var parent = (RadialContainer)Parent;
+        var isInRadius = dist < parent.Radius * 2 && dist > parent.Radius / 2;
+        if (!isInRadius)
+        {
+            return false;
+        }
+
+        var dX = cX - point.X;
+        var dY = cY - point.Y;
+        var angle = dX > 0
+            ? -MathF.Atan2(dX, dY) + MathF.PI * 2
+            : -MathF.Atan2(dX, dY);
+        var isInAngle = angle > AngleSectorFrom && angle < AngleSectorTo;
+        return isInAngle;
+    }
+
     private RadialMenu? FindParentMultiLayerContainer(Control control)
     {
         foreach (var ancestor in control.GetSelfAndLogicalAncestors())
         {
-            if (ancestor is RadialMenu)
-                return ancestor as RadialMenu;
+            if (ancestor is RadialMenu menu)
+                return menu;
         }
 
         return null;
     }
+
+    /// <summary>
+    /// Draw segment between two concentrated circles from and to certain angles.
+    /// </summary>
+    /// <param name="drawingHandleScreen">Drawing handle, to which rendering should be delegated.</param>
+    /// <param name="center">Point where circle center should be.</param>
+    /// <param name="radiusInner">Radius of internal circle.</param>
+    /// <param name="radiusOuter">Radius of external circle.</param>
+    /// <param name="angleSectorFrom">Angle in radian, from which sector should start.</param>
+    /// <param name="angleSectorTo">Angle in radian, from which sector should start.</param>
+    /// <param name="color">Color for drawing.</param>
+    /// <param name="filled">Should figure be filled, or have only border.</param>
+    private void DrawBagleSector(
+        DrawingHandleScreen drawingHandleScreen,
+        Vector2 center,
+        float radiusInner,
+        float radiusOuter,
+        float angleSectorFrom,
+        float angleSectorTo,
+        Color color,
+        bool filled = true
+    )
+    {
+        const float minimalSegmentSize = MathF.Tau / 32;
+
+        var requestedSegmentSize = angleSectorTo - angleSectorFrom;
+        var segmentCount = (int)(requestedSegmentSize / minimalSegmentSize) + 1;
+
+        var bufferSize = segmentCount * 2;
+        if (_sectorPointsForDrawing == null || _sectorPointsForDrawing.Length != bufferSize)
+        {
+            _sectorPointsForDrawing ??= new Vector2[bufferSize];
```

Comment:
This looks like it's trying to cache the calculated points, but ends up calculating them every time anyway?

Reply:
it is pre-allocating array once and not on every frame - so we lessen GC pressure